crates/isograph_schema/src/validate_client_field.rs

```diff
@@ -619,43 +632,352 @@ fn get_missing_arguments_and_validate_argument_types<'a>(
+
+                Err(WithLocation::new(
+                    ValidateSchemaError::ExpectedTypeFoundScalar { expected, actual },
+                    location,
+                ))
+            }
+        },
+    }
+}
+
+fn variable_type_satisfies_argument_type(
+    variable_type: &GraphQLTypeAnnotation<SelectableServerFieldId>,
+    argument_type: &GraphQLTypeAnnotation<SelectableServerFieldId>,
+    schema_data: &ServerFieldData,
+    location: Location,
+) -> Result<(), WithLocation<ValidateSchemaError>> {
+    match argument_type {
+        GraphQLTypeAnnotation::List(list_type) => {
+            match graphql_type_to_non_null_type(variable_type.clone()) {
+                GraphQLNonNullTypeAnnotation::List(list_variable_type) => {
+                    // [Value]! satisfies [Value]
+                    // or [Value] satisfies [Value]
+                    return variable_type_satisfies_argument_type(
+                        &list_variable_type,
+                        &list_type.0,
+                        schema_data,
+                        location,
+                    );
+                }
+                GraphQLNonNullTypeAnnotation::Named(_) => (),
+            }
+        }
+
+        GraphQLTypeAnnotation::Named(named_type) => {
+            match graphql_type_to_non_null_type(variable_type.clone()) {
+                GraphQLNonNullTypeAnnotation::Named(named_variable_type) => {
+                    // Value! satisfies Value
+                    // or Value satisfies Value
+                    if named_variable_type.item == named_type.item {
+                        return Ok(());
+                    }
+                }
+                GraphQLNonNullTypeAnnotation::List(_) => (),
+            }
+        }
+        GraphQLTypeAnnotation::NonNull(non_null_argument_type) => match variable_type {
+            // Value! satisfies Value!
+            GraphQLTypeAnnotation::NonNull(variable_type) => {
+                return variable_type_satisfies_argument_type(
+                    &graphql_type_to_nullable_type(*variable_type.clone()),
+                    &graphql_type_to_nullable_type(*non_null_argument_type.clone()),
+                    schema_data,
+                    location,
+                );
+            }
+            // Value does not satisfy Value!
+            // or [Value] does not satisfy Value!
+            GraphQLTypeAnnotation::Named(_) | GraphQLTypeAnnotation::List(_) => (),
+        },
+    };
+
+    let expected = id_annotation_to_typename_annotation(argument_type, schema_data);
+    let actual = id_annotation_to_typename_annotation(variable_type, schema_data);
+
+    Err(WithLocation::new(
+        ValidateSchemaError::ExpectedType { expected, actual },
+        location,
     ))
 }
 
-pub fn get_missing_arguments<'a>(
-    argument_definitions: impl Iterator<Item = &'a ValidatedVariableDefinition> + 'a,
+fn value_satisfies_type(
+    value: &WithLocation<NonConstantValue>,
```

Comment:
nit: `selection_supplied_argument_value` and `field_argument_definition_type` would be clearer